File: src/object_tracking_viz/src/image_publisher.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/image.hpp>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>


class ImageNode : public rclcpp::Node {
public:
    ImageNode() : Node("image_publisher"), img_count_(0) {
        RCLCPP_INFO(this->get_logger(), "Publishing Images...");
        
        std::string video_path = "/home/shivani/ros2_ws/src/object_tracking_viz/media/test1.mp4";
        camera_.open(0, cv::CAP_V4L2);
        // camera_.open(video_path);
        if (!camera_.isOpened()) {
            RCLCPP_ERROR(this->get_logger(), "Failed to open camera");
            rclcpp::shutdown();
            return;
        }

        camera_.set(cv::CAP_PROP_FOURCC, cv::VideoWriter::fourcc('M', 'J', 'P', 'G'));

        img_pub_ = this->create_publisher<sensor_msgs::msg::Image>("/camera/image_raw", 10);
        timer_ = this->create_wall_timer(std::chrono::milliseconds(40), 
                                        std::bind(&ImageNode::timer_callback, this));
    }

private:
    void timer_callback() {
        cv::Mat frame;
        if (camera_.read(frame)) {
            cv::resize(frame, frame, cv::Size(820, 620), 0, 0, cv::INTER_CUBIC);
            
            auto img_msg = cv_bridge::CvImage(std_msgs::msg::Header(), "bgr8", frame).toImageMsg();
            img_pub_->publish(*img_msg);
            RCLCPP_INFO(this->get_logger(), "Published image %d", img_count_++);
        }
    }

    rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr img_pub_;
    rclcpp::TimerBase::SharedPtr timer_;
    cv::VideoCapture camera_;
    int img_count_;
};


int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<ImageNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/object_tracking/object_tracking/image_publisher.py
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import cv2


class ImageNode(Node):
    def __init__(self):
        super().__init__('image_publisher')
        self.get_logger().info("Publishing Images...")

        self.camera = cv2.VideoCapture(0)
        self.camera.set(cv2.CAP_PROP_FOURCC, cv2.VideoWriter_fourcc('M', 'J', 'P', 'G'))
        self.img_pub = self.create_publisher(Image, '/camera/image_raw', 10)
        self.br = CvBridge()

        self.img_count = 0
        self.timer = self.create_timer(0.02, self.timer_callback)
    
    def timer_callback(self):
        success, frame = self.camera.read()
        if success:
            frame = cv2.resize(frame, (820, 620), interpolation=cv2.INTER_CUBIC)
            ros_frame = self.br.cv2_to_imgmsg(frame, "bgr8")
            self.img_pub.publish(ros_frame)
            self.get_logger().info(f"Published image {self.img_count}")
            self.img_count += 1
        

def main(args=None):
    rclpy.init(args=args)
    node = ImageNode()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()
